esi_leap/api/controllers/v1/event.py

```diff
@@ -0,0 +1,113 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import datetime
+import pecan
+from pecan import rest
+import wsme
+from wsme import types as wtypes
+import wsmeext.pecan as wsme_pecan
+
+from esi_leap.api.controllers import base
+from esi_leap.api.controllers import types
+from esi_leap.api.controllers.v1 import utils
+from esi_leap.common import exception
+from esi_leap.common import keystone
+import esi_leap.conf
+from esi_leap.objects import event as event_obj
+from esi_leap.resource_objects import get_resource_object
+
+CONF = esi_leap.conf.CONF
+
+
+class Event(base.ESILEAPBase):
+
+    id = wsme.wsattr(int, readonly=True)
+    event_type = wsme.wsattr(wtypes.text, readonly=True)
+    event_time = wsme.wsattr(datetime.datetime, readonly=True)
+    object_type = wsme.wsattr(wtypes.text, readonly=True)
+    object_uuid = wsme.wsattr(wtypes.text, readonly=True)
+    resource_type = wsme.wsattr(wtypes.text, readonly=True)
+    resource_uuid = wsme.wsattr(wtypes.text, readonly=True)
+    lessee_id = wsme.wsattr(wtypes.text, readonly=True)
+    owner_id = wsme.wsattr(wtypes.text, readonly=True)
+
+    def __init__(self, **kwargs):
+
+        self.fields = event_obj.Event.fields
+        for field in self.fields:
+            setattr(self, field, kwargs.get(field, wtypes.Unset))
+
+
+class EventCollection(types.Collection):
+    events = [Event]
+
+    def __init__(self, **kwargs):
+        self._type = 'events'
+
+
+class EventsController(rest.RestController):
+
+    @wsme_pecan.wsexpose(EventCollection, int, wtypes.text,
+                         datetime.datetime, wtypes.text, wtypes.text,
+                         wtypes.text, wtypes.text)
+    def get_all(self, last_event_id=None, lessee_or_owner_id=None,
+                last_event_time=None, event_type=None,
+                resource_type=None, resource_uuid=None):
+        request = pecan.request.context
+        cdict = request.to_policy_values()
+
+        try:
+            utils.policy_authorize('esi_leap:offer:offer_admin', cdict, cdict)
```

Comment:
I'm a little bit confused: why it checks policy esi_leap:offer:offer_admin? And who can make call to event get_all()? Admin, lessee and owner?

Reply:
Ah, right - basically, anyone can make a request. However, if the person requesting a list of events is *not* an admin, then they must be restricted to themselves. So this snippet says: if the user is not an admin, then force the query to be limited to their own project; otherwise, it's okay for the user to specify whatever project they want